[jit] Make classtypes hold a weak_ptr to their CU

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22727 [jit] Make `load()` create only one CU
* **#22726 [jit] Make classtypes hold a weak_ptr to their CU**
* #22725 [jit] Make traced fns also go into the global python CU
* #22724 [jit] _script_compile and _script_class_compile add to the python CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

Differential Revision: [D16197608](https://our.internmc.facebook.com/intern/diff/D16197608)

suo committed:
[jit] Make classtypes hold a weak_ptr to their CU

This breaks the ownership cycle between modules and their functions.

suo commented on t:
```diff
@@ -0,0 +1,58 @@
+[1mdiff --cc torch/csrc/jit/import_source.cpp[m
```
what in the world

suo committed:
Update on "[jit] Make classtypes hold a weak_ptr to their CU"

[jit] Make classtypes hold a weak_ptr to their CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

suo committed:
Update on "[jit] Make classtypes hold a weak_ptr to their CU"

[jit] Make classtypes hold a weak_ptr to their CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

suo committed:
Update on "[jit] Make classtypes hold a weak_ptr to their CU"

[jit] Make classtypes hold a weak_ptr to their CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -9,19 +9,21 @@ namespace c10 {
 }
 
-std::shared_ptr<CompilationUnit> ClassType::compilation_unit() {
+std::weak_ptr<CompilationUnit> ClassType::compilation_unit() {
```
I think this should return a strong pointer via lock(). weak_ptr is an implementation detail of the class.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -1074,8 +1074,10 @@ RegisterOperators reg(
            const auto type = node->output()->type()->expect<ClassType>();
            const size_t numAttrs = type->numAttributes();
            return [type, numAttrs](Stack& stack) {
+             auto cu = type->compilation_unit().lock();
```
StrongTypePtr and the cu lookup should go _outside_ the closure to prevent recompute every time.

zdevito reviewed (approved):
Looks good. minor nits

suo committed:
Update on "[jit] Make classtypes hold a weak_ptr to their CU"

[jit] Make classtypes hold a weak_ptr to their CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

bddppq commented:
This has broken CUDA CI on master:
```
Jul 15 21:32:01 ======================================================================
Jul 15 21:32:01 FAIL: test_lerp_cuda (__main__.TestFuser)
Jul 15 21:32:01 ----------------------------------------------------------------------
Jul 15 21:32:01 Traceback (most recent call last):
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 399, in wrapper
Jul 15 21:32:01     method(*args, **kwargs)
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 344, in __exit__
Jul 15 21:32:01     self.name, after - before, i))
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 578, in assertEqual
Jul 15 21:32:01     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
Jul 15 21:32:01 AssertionError: 512 not less than or equal to 1e-05 : __main__.TestFuser.test_lerp_cuda leaked 512 bytes CUDA memory on device 0
Jul 15 21:32:01 
Jul 15 21:32:01 ----------------------------------------------------------------------
Jul 15 21:32:01 Ran 46 tests in 28.454s
Jul 15 21:32:01 
Jul 15 21:32:01 FAILED (failures=1, skipped=2)
Jul 15 21:32:02 Traceback (most recent call last):
Jul 15 21:32:02   File "test/run_test.py", line 434, in <module>
Jul 15 21:32:02     main()
Jul 15 21:32:02   File "test/run_test.py", line 426, in main
Jul 15 21:32:02     raise RuntimeError(message)
Jul 15 21:32:02 RuntimeError: test_jit_fuser failed!
```
